pkg/store/cache/caching_bucket.go

```diff
@@ -0,0 +1,389 @@
+		res.bucket = ib.WithExpectedErrs(expectedFunc)
+		return res
+	}
+
+	// nothing else to do (?)
+	return cb
+}
+
+func (cb *CachingBucket) ReaderWithExpectedErrs(expectedFunc objstore.IsOpFailureExpectedFunc) objstore.BucketReader {
+	return cb.WithExpectedErrs(expectedFunc)
+}
+
+func (cb *CachingBucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	return cb.bucket.Iter(ctx, dir, f)
+}
+
+func (cb *CachingBucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	return cb.bucket.Get(ctx, name)
+}
+
+var chunksMatcher = regexp.MustCompile(`^.*/chunks/\d+$`)
+
+func isTSDBChunkFile(name string) bool {
+	return chunksMatcher.MatchString(name)
+}
+
+func (cb *CachingBucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	if isTSDBChunkFile(name) && off >= 0 && length > 0 {
+		return cb.getRangeChunkFile(ctx, name, off, length)
+	}
+
+	return cb.bucket.GetRange(ctx, name, off, length)
+}
+
+func (cb *CachingBucket) Exists(ctx context.Context, name string) (bool, error) {
+	return cb.bucket.Exists(ctx, name)
+}
+
+func (cb *CachingBucket) IsObjNotFoundErr(err error) bool {
+	return cb.bucket.IsObjNotFoundErr(err)
+}
+
+func (cb *CachingBucket) ObjectSize(ctx context.Context, name string) (uint64, error) {
+	return cb.bucket.ObjectSize(ctx, name)
+}
+
+func (cb *CachingBucket) cachedObjectSize(ctx context.Context, name string, ttl time.Duration) (uint64, error) {
+	key := cachingKeyObjectSize(name)
+
+	hits, _ := cb.cache.Fetch(ctx, []string{key})
+	if s := hits[key]; len(s) == 8 {
+		return binary.BigEndian.Uint64(s), nil
+	}
+
+	size, err := cb.bucket.ObjectSize(ctx, name)
+	if err != nil {
+		return 0, err
+	}
+
+	var buf [8]byte
+	binary.BigEndian.PutUint64(buf[:], size)
+	cb.cache.Store(ctx, map[string][]byte{key: buf[:]}, ttl)
+
+	return size, nil
+}
+
+func (cb *CachingBucket) getRangeChunkFile(ctx context.Context, name string, offset, length int64) (io.ReadCloser, error) {
+	size, err := cb.cachedObjectSize(ctx, name, cb.config.ChunkObjectSizeTTL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to get size of chunk file: %s", name)
+	}
+
+	// If length goes over object size, adjust length. We use it later to limit number of read bytes.
+	if uint64(offset+length) > size {
+		length = int64(size - uint64(offset))
+	}
+
+	startOffset := (offset / cb.config.ChunkBlockSize) * cb.config.ChunkBlockSize
+	endOffset := ((offset + length) / cb.config.ChunkBlockSize) * cb.config.ChunkBlockSize
+	if (offset+length)%cb.config.ChunkBlockSize > 0 {
+		endOffset += cb.config.ChunkBlockSize
+	}
+
+	blocks := (endOffset - startOffset) / cb.config.ChunkBlockSize
+
+	offsetKeys := make(map[int64]string, blocks)
+	keys := make([]string, 0, blocks)
+	lastBlockLength := 0
+
+	for o := startOffset; o < endOffset; o += cb.config.ChunkBlockSize {
+		end := o + cb.config.ChunkBlockSize
+		if end > int64(size) {
+			end = int64(size)
+		}
+
+		lastBlockLength = int(end - o)
```

Comment:
Despite correct, looks a bit weird updating the `lastBlockLength` at every `for` iteration when actually it's just the last iteration the correct one. You can compute the `lastBlockLength` in a 1-line when you define the variable above. Have you considered doing it?